# BATTLE OF NEIGHBOURHOODS - TORONTO

In this notebook we will explore, segment, and cluster the neighborhoods in the city of Toronto based on the postalcode and borough information

In [3]:
!pip install -U numpy

!pip install -U pandas

!pip install -U scipy

!pip install -U scikit-learn

!pip install -U imbalanced-learn

     |████████████████████████████████| 27.4 MB 13.2 MB/s eta 0:00:01
  Attempting uninstall: scipy
    Found existing installation: scipy 1.5.3
    Uninstalling scipy-1.5.3:
      Successfully uninstalled scipy-1.5.3
     |████████████████████████████████| 206 kB 13.8 MB/s eta 0:00:01


In [1]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [2]:
import pandas as pd

# QUESTION 1 

Load the data and then create a data frame which consist of three columns:PostalCode, Borough, and Neighborhood

## Importing data

In [3]:
import requests

In [4]:
url="https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=1011037969"
wiki_url=requests.get(url)
wiki_url

<Response [200]>

here,response 200 means that we are able to make connection to the page 

In [5]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

As we only need the first table so ,we'll drop the other tables.

In [6]:
wiki_data=wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [7]:
wiki_data.shape

(180, 3)

Next step , we have to drop borough which is not assigned

In [8]:
df=wiki_data[wiki_data["Borough"]!='Not assigned']
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Now, grouping by postal code

In [9]:
df= df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Checking for number of records where Neighbourhood is "Not assigned"

In [10]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [11]:
df.shape

(103, 3)

## Answer to Question 1: We have 130 rows and 3 columns

# QUESTION 2

Installing geocoder

In [12]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


In [13]:
import geocoder #importing geocoder

Already tried the given approach in the assignment ,it ran for 20 min then killed it. Alternatively, as suggested in the assignment, Importing the CSV file from the URL

In [14]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [15]:
print("The shape of wiki_data is {}".format(df.shape))
print("The shape of wiki_data is {}".format(data.shape))

The shape of wiki_data is (103, 3)
The shape of wiki_data is (103, 3)


As,the dimension of both data sets are equal

we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [16]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [17]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [18]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
2                                            Parkwoods  43.753259 -79.329656  
3                                     Victoria Village  43.725882 -79.315572  
4                            Regent Park, Harbourfront  43.654260 -79.360636  
5                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
6          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
160      The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
165                               Church and Wellesley  43.665860 -79.383160  
168  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
178  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [19]:
combined_data.shape

(103, 5)

## Answer 2

We get 103 rows as expected when we do a inner join, so we have good data.

# QUESTION 3

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [20]:
from geopy.geocoders import Nominatim

In [21]:
address ='Toronto, Ontario'
geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [22]:

import folium
print("folium imported")

folium imported


In [23]:
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(map_Toronto)
map_Toronto

Initializing Foursquare API credentials

In [30]:
CLIENT_ID = 'T5JOJMUDDRZO3UU1W1Z54OZM2IACDHXOXWWHH1CU4CXPV0GN' 
CLIENT_SECRET = 'UNJN41PZFOD5LLAYIZAKQN0CORNRS3OHIWL51TFVRCGGOYIP'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: T5JOJMUDDRZO3UU1W1Z54OZM2IACDHXOXWWHH1CU4CXPV0GN
CLIENT_SECRET:UNJN41PZFOD5LLAYIZAKQN0CORNRS3OHIWL51TFVRCGGOYIP


Next, we create a function to get all the venue categories in Toronto

In [31]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)


In [32]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [33]:
venues_in_toronto.shape

(1323, 5)

So we have 1323 records and 5 columns. Checking sample data

In [34]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3               Portugril  Portuguese Restaurant  
4             Tim Hortons            Coffee Shop

In [35]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1309  Mimico NW, The Queensway West, South of Bloor,...   
1310  Mimico NW, The Queensway West, South of Bloor,...   
1311  Mimico NW, The Queensway West, South of Bloor,...   
1312  Mimico NW, The Queensway West, South of Bloor,...   
1313  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572               Portugril   
4                  43.725882               -79.315572             Tim Hortons   
...                      ...                      ...                     ...   
1309               43.628841               -79.520999              Wingporium   
1310               43.628841               -79.520999        South St. Burger   
1311               43.628841               -79.520999               Dollarama   
1312               43.628841               -79.520999          Healthy Planet   
1313               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3     Portuguese Restaurant  
4               Coffee Shop  
...                     ...  
1309            Wings Joint  
1310           Burger Joint  
1311         Discount Store  
1312        Supplement Shop  
1313                 Bakery  

[411 rows x 5 columns]

So there are 405 records for each neighbourhood.

Checking for the maximum venue categories

In [36]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
Airport Service     CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint               Queen's Park, Ontario Provincial Government   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
Airport Service                  43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.400049   
Wings Joint                      43.662301               -79.389494   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Lounge                        Porter Lounge  
Airport Service            ORNGE - Toronto Air Base  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[231 rows x 4 columns]

# One Hot encoding the venue Categories

In [37]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Lounge  \
0                     0        0                   0               0   
1                     0        0                   0               0   
2                     0        0                   0               0   
3                     0        0                   0               0   
4                     0        0                   0               0   
...                 ...      ...                 ...             ...   
1318                  0        0                   0               0   
1319                  0        0                   0               0   
1320                  0        0                   0               0   
1321                  0        0                   0               0   
1322                  0        0                   0               0   

      Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                   0                 0                    0             0   
1                   0                 0                    0             0   
2                   0                 0                    0             0   
3                   0                 0                    0             0   
4                   0                 0                    0             0   
...               ...               ...                  ...           ...   
1318                0                 0                    0             0   
1319                0                 0                    0             0   
1320                0                 0                    0             0   
1321                0                 0                    0             0   
1322                0                 0                    0             0   

      Aquarium  Art Gallery  ...  Train Station  Truck Stop  \
0            0            0  ...              0           0   
1            0            0  ...              0           0   
2            0            0  ...              0           0   
3            0            0  ...              0           0   
4            0            0  ...              0           0   
...        ...          ...  ...            ...         ...   
1318         0            0  ...              0           0   
1319         0            0  ...              0           0   
1320         0            0  ...              0           0   
1321         0            0  ...              0           0   
1322         0            0  ...              0           0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1318                              0                 0                      0   
1319                              0                 0                      0   
1320                              0                 0                      0   
1321                              0                 0                      0   
1322                              0                 0                      0   

      Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                   0         0            0              0            0  
1                   0         0            0              0            0  
2                   0         0            0              0            0  
3                   0         0            0              0            0  
4                   0         0            0              0            0  
...               ...       ...     

In [38]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()


Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Train Station  Truck Stop  \
0             0         0  ...              0           0   
1             0         0  ...              0           0   
2             0         0  ...              0           0   
3             0         0  ...              0           0   
4             0         0  ...              0           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 232 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood

In [39]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Lounge  Airport Service  \
0      0.0                 0.0             0.0              0.0   
1      0.0                 0.0             0.0              0.0   
2      0.0                 0.0             0.0              0.0   
3      0.0                 0.0             0.0              0.0   
4      0.0                 0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  ...  \
0               0.0             0.000000           0.0       0.0  ...   
1               0.0             0.000000           0.0       0.0  ...   
2               0.0             0.000000           0.0       0.0  ...   
3               0.0             0.000000           0.0       0.0  ...   
4               0.0             0.038462           0.0       0.0  ...   

   Train Station  Truck Stop  Vegetarian / Vegan Restaurant  Video Game Store  \
0            0.0         0.0                            0.0               0.0   
1            0.0         0.0                            0.0               0.0   
2            0.0         0.0                            0.0               0.0   
3            0.0         0.0                            0.0               0.0   
4            0.0         0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 232 columns]


Let's make a function to get the top most common venue categories

In [40]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [41]:
import numpy as np


In [42]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village    Chinese Restaurant   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant        Clothing Store   
1           Coffee Shop               Skating Rink        Sandwich Place   
2           Coffee Shop                Pizza Place          Intersection   
3                  Bank                       Café   Japanese Restaurant   
4        Sandwich Place                  Juice Bar    Italian Restaurant   

  5th Most Common Venue      6th Most Common Venue     7th Most Common Venue  \
0     Accessories Store              Movie Theater  Mediterranean Restaurant   
1                   Pub                        Gym                Playground   
2           Bridal Shop  Middle Eastern Restaurant                  Pharmacy   
3     Accessories Store              Movie Theater            Medical Center   
4         Grocery Store            Thai Restaurant                  Pharmacy   

      8th Most Common Venue     9th Most Common Venue  \
0               Men's Store        Mexican Restaurant   
1            Medical Center  Mediterranean Restaurant   
2             Shopping Mall         Mobile Phone Shop   
3  Mediterranean Restaurant               Men's Store   
4   Comfort Food Restaurant                       Pub   

      10th Most Common Venue  
0  Middle Eastern Restaurant  
1             Massage Studio  
2                Supermarket  
3         Mexican Restaurant  
4                 Restaurant

## Let's make the model to cluster our Neighbourhoods

In [43]:
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

Checking the labelling of our model




In [44]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 4, 1, 1, 1, 4, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 1,
       1, 1, 4, 1, 1, 0, 4, 1, 4, 1, 1, 1, 1, 1, 3, 1, 1, 4, 1, 1, 1, 4,
       2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4,
       1, 4, 1, 1, 1, 1, 4], dtype=int32)

Let's add the clustering Label column to the top 10 common venue categories

In [45]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [47]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
2         M3A        North York                                    Parkwoods   
3         M4A        North York                             Victoria Village   
4         M5A  Downtown Toronto                    Regent Park, Harbourfront   
5         M6A        North York             Lawrence Manor, Lawrence Heights   
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
2  43.753259 -79.329656             4.0                  Park   
3  43.725882 -79.315572             1.0     French Restaurant   
4  43.654260 -79.360636             1.0           Coffee Shop   
5  43.718518 -79.464763             1.0        Clothing Store   
6  43.662301 -79.389494             1.0           Coffee Shop   

   2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
2      Food & Drink Shop      Accessories Store                 Motel   
3  Portuguese Restaurant            Coffee Shop          Hockey Arena   
4                   Café                   Park                Bakery   
5      Accessories Store  Vietnamese Restaurant    Miscellaneous Shop   
6       Sushi Restaurant            Yoga Studio   Distribution Center   

  5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
2   Martial Arts School        Massage Studio            Medical Center   
3     Accessories Store        Medical Center  Mediterranean Restaurant   
4        Breakfast Spot           Yoga Studio              Dessert Shop   
5              Boutique           Coffee Shop    Furniture / Home Store   
6    Mexican Restaurant         Burrito Place                      Café   

      8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
2  Mediterranean Restaurant           Men's Store         Mexican Restaurant  
3               Men's Store    Mexican Restaurant  Middle Eastern Restaurant  
4            Farmers Market     French Restaurant        Distribution Center  
5       Sporting Goods Shop   Monument / Landmark        Martial Arts School  
6                  Creperie        Sandwich Place        Fried Chicken Joint


Drop all the NaN values to prevent data skew

In [48]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

In [49]:
import matplotlib.cm as cm
import matplotlib.colors as colors


In [50]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters